### PR TITLE
feat: add batch replicate from remote MinIO to local cluster

### PR DESCRIPTION
**Pull request by poornas**
between 2 minio instances

## Motivation and Context
This option may be useful when remote cluster is on older release that does not support batch jobs yet

## How to test this PR?
```
replicate:
  apiVersion: v1
  # source of the objects to be replicated
  source:
    type: minio # valid values are "minio"
    bucket: bucket
    prefix:  # 'PREFIX' is optional
    # NOTE: if source is remote then target must be "local"
    endpoint: http://localhost:9004
    credentials:
      accessKey: minio
      secretKey: minio123
  #  #   sessionToken: SESSION-TOKEN # Optional only available when rotating credentials are used

  # target where the objects must be replicated
  target:
    type: minio # valid values are "minio"
    bucket: bucketa
    prefix: xxx # 'PREFIX' is optional
    # NOTE: if target is remote then source must be "local"
    # endpoint: http://localhost:9004
    # credentials:
    #   accessKey: minio
    #   secretKey: minio123
    # #   sessionToken: SESSION-TOKEN # Optional only available when rotating credentials are used
...
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/batch-handlers.go [hunk 1]
```diff
@@ -1154,13 +1548,24 @@ func (j *BatchJobPool) AddWorker() {
 				return
 			}
 			if job.Replicate != nil {
-				if err := job.Replicate.Start(job.ctx, j.objLayer, *job); err != nil {
-					if !isErrBucketNotFound(err) {
-						logger.LogIf(j.ctx, err)
-						j.canceler(job.ID, false)
-						continue
+				if job.Replicate.RemoteToLocal() {
+					if err := job.Replicate.startReplFromSource(job.ctx, j.objLayer, *job); err != nil {
```
Comment (harshavardhana) [inline cmd/batch-handlers.go]:
```suggestion
					if err := job.Replicate.StartFromSource(job.ctx, j.objLayer, *job); err != nil {
```

Comment (harshavardhana) [reply cmd/batch-handlers.go]: Replicate is already the parent variable.

File: cmd/batch-handlers.go [hunk 2]
```diff
@@ -270,10 +285,360 @@ func (r BatchJobReplicateV1) Notify(ctx context.Context, body io.Reader) error {
...
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			// block and abort remote upload upon failure.
+			attempts := 1
+			for attempts <= 3 {
+				aerr := c.AbortMultipartUpload(ctx, tgtBucket, tgtObject, res.UploadID)
```
Comment (harshavardhana) [inline cmd/batch-handlers.go]: should the multipart abort be local not remote? @poornas

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Rest if the changes look good

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
Is the source type always MinIO currently?

**Comment by poornas** [comment]
> Is the source type always MinIO currently?

yes

**Pull request merged**